// @wordpress/block-editor/build-module/components/block-navigation/block-select-button.js
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { __experimentalGetBlockLabel as getBlockLabel, getBlockType } from '@wordpress/blocks';
import { Button, VisuallyHidden } from '@wordpress/components';
import { useInstanceId } from '@wordpress/compose';
import { forwardRef } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import BlockIcon from '../block-icon';
import Indentation from './indentation';
import useBlockDisplayInformation from '../use-block-display-information';
import { getBlockPositionDescription } from './utils';

function BlockNavigationBlockSelectButton(_ref, ref) {
  var className = _ref.className,
      _ref$block = _ref.block,
      clientId = _ref$block.clientId,
      name = _ref$block.name,
      attributes = _ref$block.attributes,
      isSelected = _ref.isSelected,
      onClick = _ref.onClick,
      position = _ref.position,
      siblingBlockCount = _ref.siblingBlockCount,
      level = _ref.level,
      tabIndex = _ref.tabIndex,
      onFocus = _ref.onFocus,
      onDragStart = _ref.onDragStart,
      onDragEnd = _ref.onDragEnd,
      draggable = _ref.draggable;
  var blockInformation = useBlockDisplayInformation(clientId);
  var instanceId = useInstanceId(BlockNavigationBlockSelectButton);
  var descriptionId = "block-navigation-block-select-button__".concat(instanceId);
  var blockType = getBlockType(name);
  var blockLabel = getBlockLabel(blockType, attributes); // If label is defined we prioritize it over possible possible
  // block variation match title.

  var blockDisplayName = blockLabel !== blockType.title ? blockLabel : blockInformation === null || blockInformation === void 0 ? void 0 : blockInformation.title;
  var blockPositionDescription = getBlockPositionDescription(position, siblingBlockCount, level);
  return createElement(Fragment, null, createElement(Button, {
    className: classnames('block-editor-block-navigation-block-select-button', className),
    onClick: onClick,
    "aria-describedby": descriptionId,
    ref: ref,
    tabIndex: tabIndex,
    onFocus: onFocus,
    onDragStart: onDragStart,
    onDragEnd: onDragEnd,
    draggable: draggable
  }, createElement(Indentation, {
    level: level
  }), createElement(BlockIcon, {
    icon: blockInformation === null || blockInformation === void 0 ? void 0 : blockInformation.icon,
    showColors: true
  }), blockDisplayName, isSelected && createElement(VisuallyHidden, null, __('(selected block)'))), createElement("div", {
    className: "block-editor-block-navigation-block-select-button__description",
    id: descriptionId
  }, blockPositionDescription));
}

export default forwardRef(BlockNavigationBlockSelectButton);
